File: windows/TrikiMusicController.Windows/Core/EdgePoseBrightnessController.cs
using TrikiMusicController_Windows.Models;

namespace TrikiMusicController_Windows.Core;

public sealed record BrightnessControlResult(
    bool Active,
    bool Ready,
    float BrightnessPercent,
    float DeltaPercent,
    float StabilizationProgress,
    string StatusText);

public sealed class EdgePoseBrightnessController
{
    public const long DefaultStabilizationNanos = 150_000_000; // 150 ms szybkiej stabilizacji
    public const float DegreesPerPercentBrightness = 2.5f; // 250 deg = 100% jasności
    public const float GyroscopeDeadbandDps = 3.0f;
    public const float EdgeEnterMaxZ = 0.45f;
    public const float EdgeExitMaxZ = 0.60f;
    public const float EdgeMinPlaneG = 0.65f;
    public const float EdgeMaxPlaneG = 1.35f;

    private readonly long _stabilizationNanos;
    private long? _stabilizationStartNanos;
    private long? _lastTimestampNanos;
    private float _accumulatedDegrees;
    private float _currentBrightnessPercent = 60f;
    private bool _isCurrentlyInEdge;

    public EdgePoseBrightnessController(
        float initialBrightnessPercent = 60f,
        long stabilizationNanos = DefaultStabilizationNanos)
    {
        _currentBrightnessPercent = Math.Clamp(initialBrightnessPercent, 0f, 100f);
        _stabilizationNanos = stabilizationNanos;
    }

    public float CurrentBrightnessPercent
    {
        get => _currentBrightnessPercent;
        set => _currentBrightnessPercent = Math.Clamp(value, 0f, 100f);
    }

    public void Reset()
    {
        _stabilizationStartNanos = null;
        _lastTimestampNanos = null;
        _accumulatedDegrees = 0f;
        _isCurrentlyInEdge = false;
    }

    public BrightnessControlResult Process(FilteredSensorData sample, bool isButtonPressed = true)
    {
        var accZ = Math.Abs(sample.AccelerometerG.Z);
        var planeAcc = MathF.Sqrt(
            sample.AccelerometerG.X * sample.AccelerometerG.X +
            sample.AccelerometerG.Y * sample.AccelerometerG.Y);

        // Histereza pozycji krawędziowej 90°
        var isEdgePose = _isCurrentlyInEdge
            ? (accZ <= EdgeExitMaxZ && planeAcc >= EdgeMinPlaneG * 0.8f)
            : (accZ <= EdgeEnterMaxZ && planeAcc >= EdgeMinPlaneG && planeAcc <= EdgeMaxPlaneG);

        _isCurrentlyInEdge = isEdgePose;
        var timestamp = sample.Source.TimestampNanos;

        if (!isEdgePose)
        {
            _stabilizationStartNanos = null;
            _lastTimestampNanos = timestamp;
            _accumulatedDegrees = 0f;
            return new BrightnessControlResult(
                Active: false,
                Ready: false,
                BrightnessPercent: _currentBrightnessPercent,
                DeltaPercent: 0f,
                StabilizationProgress: 0f,
                StatusText: "Postaw kapsel na krawędzi (90°) i przytrzymaj przycisk, aby regulować jasność.");
        }

        if (_stabilizationStartNanos is null)
        {
            _stabilizationStartNanos = timestamp;
            _lastTimestampNanos = timestamp;
            return new BrightnessControlResult(
                Active: true,
                Ready: false,
                BrightnessPercent: _currentBrightnessPercent,
                DeltaPercent: 0f,
                StabilizationProgress: 0f,
                StatusText: "Stabilizacja pozycji 90°…");
        }

        var elapsedStabilization = timestamp - _stabilizationStartNanos.Value;
        var stabilizationProgress = Math.Clamp((float)elapsedStabilization / _stabilizationNanos, 0f, 1f);

        if (elapsedStabilization < _stabilizationNanos && !isButtonPressed)
        {
            _lastTimestampNanos = timestamp;
            return new BrightnessControlResult(
                Active: true,
                Ready: false,
                BrightnessPercent: _currentBrightnessPercent,
                DeltaPercent: 0f,
                StabilizationProgress: stabilizationProgress,
                StatusText: $"Stabilizacja: {(int)(stabilizationProgress * 100)}%");
        }

        if (!isButtonPressed)
        {
            _accumulatedDegrees = 0f;
            _lastTimestampNanos = timestamp;
            return new BrightnessControlResult(
                Active: true,
                Ready: false,
                BrightnessPercent: _currentBrightnessPercent,
                DeltaPercent: 0f,
                StabilizationProgress: 1f,
                StatusText: "Przytrzymaj przycisk, aby regulować jasność w pozycji 90°.");
        }

        var deltaPercent = 0f;
        if (_lastTimestampNanos is long prevTimestamp && timestamp > prevTimestamp)
        {
            var deltaNanos = timestamp - prevTimestamp;
            var dtSeconds = (deltaNanos > 250_000_000L) ? 0.02f : (deltaNanos / 1_000_000_000f);
            var gyroZ = sample.GyroscopeDps.Z;

            if (Math.Abs(gyroZ) >= GyroscopeDeadbandDps)
            {
                var deltaDegrees = gyroZ * dtSeconds;
                _accumulatedDegrees += deltaDegrees;

                if (Math.Abs(_accumulatedDegrees) >= DegreesPerPercentBrightness)
                {
                    var steps = MathF.Truncate(_accumulatedDegrees / DegreesPerPercentBrightness);
                    deltaPercent = steps;
                    _accumulatedDegrees -= steps * DegreesPerPercentBrightness;
                    _currentBrightnessPercent = Math.Clamp(_currentBrightnessPercent + deltaPercent, 0f, 100f);
                }
            }
        }

        _lastTimestampNanos = timestamp;

        return new BrightnessControlResult(
            Active: true,
            Ready: true,
            BrightnessPercent: _currentBrightnessPercent,
            DeltaPercent: deltaPercent,
            StabilizationProgress: 1f,
            StatusText: $"Jasność: {(int)_currentBrightnessPercent}% (Obracaj trzymając przycisk)");
    }
}

File: app/src/main/java/pl/trikimusic/controller/core/brightness/EdgePoseBrightnessController.kt
package pl.trikimusic.controller.core.brightness

import kotlin.math.abs
import kotlin.math.sqrt
import pl.trikimusic.controller.domain.model.FilteredSensorData

data class BrightnessControlResult(
    val active: Boolean,
    val ready: Boolean,
    val brightnessPercent: Float,
    val deltaPercent: Float,
    val stabilizationProgress: Float,
    val statusText: String,
)

class EdgePoseBrightnessController(
    initialBrightnessPercent: Float = 50f,
    private val stabilizationNanos: Long = DEFAULT_STABILIZATION_NANOS,
) {
    companion object {
        const val DEFAULT_STABILIZATION_NANOS = 150_000_000L // 150 ms szybkiej stabilizacji
        const val DEGREES_PER_PERCENT_BRIGHTNESS = 2.5f // 250 deg = 100% jasności
        const val GYROSCOPE_DEADBAND_DPS = 3.0f
        const val EDGE_ENTER_MAX_Z = 0.45f
        const val EDGE_EXIT_MAX_Z = 0.60f
        const val MINIMUM_PLANE_ACCELERATION_G = 0.65f
        const val MAXIMUM_PLANE_ACCELERATION_G = 1.35f
    }

    private var currentBrightnessPercent = initialBrightnessPercent.coerceIn(0f, 100f)
    private var stabilizationStartNanos: Long? = null
    private var lastTimestampNanos: Long? = null
    private var accumulatedDegrees = 0f
    private var isCurrentlyInEdge = false

    fun getBrightness(): Float = currentBrightnessPercent

    fun setBrightness(value: Float) {
        currentBrightnessPercent = value.coerceIn(0f, 100f)
    }

    fun reset() {
        stabilizationStartNanos = null
        lastTimestampNanos = null
        accumulatedDegrees = 0f
        isCurrentlyInEdge = false
    }

    fun process(sample: FilteredSensorData, isButtonPressed: Boolean = true): BrightnessControlResult {
        val accZ = abs(sample.accelerometerG.z)
        val planeAcc = sqrt(
            sample.accelerometerG.x * sample.accelerometerG.x +
            sample.accelerometerG.y * sample.accelerometerG.y,
        )

        // Histereza pozycji krawędziowej 90°
        val isEdgePose = if (isCurrentlyInEdge) {
            accZ <= EDGE_EXIT_MAX_Z && planeAcc >= MINIMUM_PLANE_ACCELERATION_G * 0.8f
        } else {
            accZ <= EDGE_ENTER_MAX_Z && planeAcc >= MINIMUM_PLANE_ACCELERATION_G && planeAcc <= MAXIMUM_PLANE_ACCELERATION_G
        }

        isCurrentlyInEdge = isEdgePose
        val timestamp = sample.source.timestampNanos

        if (!isEdgePose) {
            stabilizationStartNanos = null
            lastTimestampNanos = timestamp
            accumulatedDegrees = 0f
            return BrightnessControlResult(
                active = false,
                ready = false,
                brightnessPercent = currentBrightnessPercent,
                deltaPercent = 0f,
                stabilizationProgress = 0f,
                statusText = "Postaw kapsel na krawędzi (90°) i przytrzymaj przycisk, aby regulować jasność.",
            )
        }

        if (stabilizationStartNanos == null) {
            stabilizationStartNanos = timestamp
            lastTimestampNanos = timestamp
            return BrightnessControlResult(
                active = true,
                ready = false,
                brightnessPercent = currentBrightnessPercent,
                deltaPercent = 0f,
                stabilizationProgress = 0f,
                statusText = "Stabilizacja pozycji 90°…",
            )
        }

        val elapsedStabilization = timestamp - stabilizationStartNanos!!
        val stabilizationProgress = (elapsedStabilization.toFloat() / stabilizationNanos).coerceIn(0f, 1f)

        if (elapsedStabilization < stabilizationNanos && !isButtonPressed) {
            lastTimestampNanos = timestamp
            return BrightnessControlResult(
                active = true,
                ready = false,
                brightnessPercent = currentBrightnessPercent,
                deltaPercent = 0f,
                stabilizationProgress = stabilizationProgress,
                statusText = "Stabilizacja: ${(stabilizationProgress * 100).toInt()}%",
            )
        }

        if (!isButtonPressed) {
            accumulatedDegrees = 0f
            lastTimestampNanos = timestamp
            return BrightnessControlResult(
                active = true,
                ready = false,
                brightnessPercent = currentBrightnessPercent,
                deltaPercent = 0f,
                stabilizationProgress = 1f,
                statusText = "Przytrzymaj przycisk, aby regulować jasność w pozycji 90°.",
            )
        }

        var deltaPercent = 0f
        val prevTimestamp = lastTimestampNanos
        if (prevTimestamp != null && timestamp > prevTimestamp) {
            val deltaNanos = timestamp - prevTimestamp
            val dtSeconds = if (deltaNanos > 250_000_000L) 0.02f else (deltaNanos / 1_000_000_000f)
            val gyroZ = sample.gyroscopeDps.z

            if (abs(gyroZ) >= GYROSCOPE_DEADBAND_DPS) {
                val deltaDegrees = gyroZ * dtSeconds
                accumulatedDegrees += deltaDegrees

                if (abs(accumulatedDegrees) >= DEGREES_PER_PERCENT_BRIGHTNESS) {
                    val steps = (accumulatedDegrees / DEGREES_PER_PERCENT_BRIGHTNESS).toInt().toFloat()
                    deltaPercent = steps
                    accumulatedDegrees -= steps * DEGREES_PER_PERCENT_BRIGHTNESS
                    currentBrightnessPercent = (currentBrightnessPercent + deltaPercent).coerceIn(0f, 100f)
                }
            }
        }

        lastTimestampNanos = timestamp

        return BrightnessControlResult(
            active = true,
            ready = true,
            brightnessPercent = currentBrightnessPercent,
            deltaPercent = deltaPercent,
            stabilizationProgress = 1f,
            statusText = "Jasność: ${currentBrightnessPercent.toInt()}% (Obracaj trzymając przycisk)",
        )
    }
}
